### PR TITLE
Dashboard: Add support for Tempo query variables

### LINKED ISSUE #70305: Template variables of type "Query" does not show editor, maybe it's not available at all
**Is your feature request related to a problem? Please describe.**
I'd like to configure query template variable with Tempo datasource ,for example to retrieve the list of service names, but I can't find a way, it seems the query variable is only available on few datasources; I can select Tempo but no query editor shows up.

**Describe the solution you'd like**
Just like with Prometheus, it would be great to use Tempo as datasource for creating query variables

**Describe alternatives you've considered**
The actual alternative is to use prometheus and fetch the services from it.

**Additional context**

### LINKED ISSUE #72575: Tempo template variables

### PR DESCRIPTION
**Pull request by fabrizio-grafana**
**What is this feature?**
Add support for template variables (`Label names` and `Label values`) in Tempo, similarly to what is already available in Loki and Prometheus. [Quick demo](https://github.com/grafana/grafana/assets/135109076/7074728c-8bf4-4251-a27b-21bcb718e2f6).

**Why do we need this feature?**
It has been requested by a few customers and discussed with @09jvilla.

**Who is this feature for?**
Customers.

**Which issue(s) does this PR fix?**:
Fixes https://github.com/grafana/grafana/issues/72575
Fixes https://github.com/grafana/grafana/issues/70305

### CODE REVIEW

**Review by 09jvilla — APPROVED** [review]
Didn't review the code but approving based on the functionality demonstrated. Amazing work!

We have discussed in a separate thread that we should create another PR that ensures Tempo query variables work when the dashboard panel query is made via the `Search` tab. Today they only work if the query is made via the `TraceQL` tab.

File: public/app/plugins/datasource/tempo/mocks.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+    access: 'direct',
+    ...settings,
+  };
+
+  // @ts-expect-error
+  return new TempoDatasource(customSettings, templateSrvMock, timeSrvStub);
+}
+
+export function createMetadataRequest(labelsAndValues: Record<string, Record<string, string[]>>) {
+  return async (url: string, params?: Record<string, string | number>) => labelsAndValues;
```
Comment (joey-grafana) [inline public/app/plugins/datasource/tempo/mocks.ts:67]: url/params unused

**Review by joey-grafana — COMMENTED** [review]
Could we modify the fields here so that there is one per row? We could also shorten the label width here as they seem quite long. In contrast, I think we should make the second fields dropdown input a bit longer. As you can see in the image, it's quite short which makes it hard to see what was selected in the dropdown.

Screenshot 2023-08-18 at 14 54 36

File: public/app/plugins/datasource/tempo/VariableQueryEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+      {type === TempoVariableQueryType.LabelValues && (
+        <InlineField label="Label" labelWidth={20}>
+          <Select
+            aria-label="Label"
+            onChange={onLabelChange}
+            onBlur={handleBlur}
+            value={{ label, value: label }}
+            options={labelOptions}
+            width={16}
+            allowCustomValue
```
[the remark this answers is not in the document]
Comment (fabrizio-grafana) [reply public/app/plugins/datasource/tempo/VariableQueryEditor.tsx:96]: Yes. And now that I think about it, I guess it doesn't make sense to allow for custom values. I'll remove this prop.

[the remark this answers is not in the document]
Comment (aocenas) [reply public/app/plugins/datasource/tempo/VariableQueryEditor.tsx]: Actualy this has to allow custom variables. Apart from what we talked about in standup that we may not know all the values in advance, the common pattern with template variables is that users chain them. So you can actually use a some different template variable here.

**Comment by aocenas** [comment]
@fabrizio-grafana did you talk with somebody regarding whether we need or don't need also a search result (so TraceIDs) as a template variable?

**Comment by 09jvilla** [comment]
> @fabrizio-grafana did you talk with somebody regarding whether we need or don't need also a search result (so TraceIDs) as a template variable?

@aocenas -- This is something we're leaving out of scope for now and will revisit in the future if there is sufficient demand.

**Review by joey-grafana — COMMENTED** [review]
This looks really good and will be a great addition :) 

I've noticed something that we should fix before merging. Nothing major so should be ok to fix. Querying the v2 endpoint always returns an error because we are not providing the scope in the query.

i.e. the query is like `api/v2/search/tag/job/values` when it should be `api/v2/search/tag/resource.job/values`. As a result, we always fallback to v1 of the API. In this case I believe that tags result are the exact same so this still works, however we should probably fix the issue with the API v2 tags query as we will more than likely eventually drop support for v1 and move to v2 + it removes one error from the network tab/console.

Screenshot 2023-08-24 at 12 27 53

**Review by joey-grafana — APPROVED** [review]
This looks great! 👍

**Pull request merged**